src/ocean/math/WideUInt.d

```diff
@@ -0,0 +1,591 @@
+/*******************************************************************************
+
+    Implementation struct emulating positive number with value range exceeding
+    one of `ulong` but still fixed in size (defined via template argument).
+
+    Such struct is a value type with stable binary layout for a given size which
+    is the primary goal of this implementation. Performance may be eventually
+    improved if necessity arises, right now implementation is intentionally
+    simplistic and naive.
+
+    Internally wide integer is represented by a static array of `uint` values
+    such that concatenating their binary representation together results in long
+    binary sequence representing actual stored number.
+
+    Copyright:
+        Copyright (c) 2017 Sociomantic Labs GmbH.
+        All rights reserved.
+
+    License:
+        Boost Software License Version 1.0. See LICENSE_BOOST.txt for details.
+        Alternatively, this file may be distributed under the terms of the Tango
+        3-Clause BSD License (see LICENSE_BSD.txt for details).
+
+*******************************************************************************/
+
+module ocean.math.WideUInt;
+
+import ocean.transition;
+import ocean.core.Test;
+import ocean.core.Enforce;
+
+import core.stdc.math;
+import ocean.math.IEEE : feqrel;
+
+/// ditto
+public struct WideUInt ( size_t N )
+{
+    static assert (
+        N > 2,
+        "For 'N <= 2' using 'uint' or 'ulong' directly is suggested"
+    );
+
+    /***************************************************************************
+
+       Binary representation of the number can be formed by putting binary
+       representation of each individual word in `payload` side by
+       side starting from the `payload[$-1]`.
+
+       `uint` is chosen as individual word type because multiplying two `uint`
+       words will result in `ulong` at most, making possible to use basic
+       integer arithmetic to implement most of wide integer arithmetic.
+
+    ***************************************************************************/
+
+    private uint[N] payload;
+
+    version (D_Version2)
```

Comment:
Nope, this is required for compatibility between D1 and D2. WideUInt num = 42 will result in opAssign being called in D1 but in D2 it requires constructor.